### ISSUE TITLE
[UX] move text filter configuration into modals

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1032]
(Related: #975)

https://www.drupal.org/project/drupal/issues/1834682

Before:

![screen shot 2018-11-02 at 6 52 37 pm](https://user-images.githubusercontent.com/2423362/47900311-8faf6300-ded0-11e8-8415-1f7b3742345d.png)

After:

![screen shot 2018-11-02 at 6 54 29 pm](https://user-images.githubusercontent.com/2423362/47900384-d43afe80-ded0-11e8-91a0-71b15bb6b277.png)

---
~~PR by @docwilmot: https://github.com/backdrop/backdrop/pull/1023~~
~~PR by @docwilmot: https://github.com/backdrop/backdrop/pull/3074~~
~PR by @docwilmot: https://github.com/backdrop/backdrop/pull/3077~
~PR by @docwilmot: https://github.com/backdrop/backdrop/pull/3096 using dialogs~
PR by @docwilmot and @quicksketch: https://github.com/backdrop/backdrop/pull/3131

**Comment by quicksketch** [comment]
@klonos, I'd be interested to hear your thoughts on https://www.drupal.org/node/1834682, where I proposed moving the filter enabling/disabling to be within the vertical tabs. That issue was held up on debate, but I still think it's worth pursuing.

**Comment by docwilmot** [comment]
:+1:

**Comment by klonos** [comment]
@quicksketch I was just about to file an issue about the very same thing just minutes before you posted your comment above. I was thinking to merge the enabling/disabling with the vtabs and -if possible- allowing tabs to be dragged/reordered. You read my mind. Will share my 2c in that issue + open one for Backdrop.

**Comment by klonos** [comment]
...in the meantime, this issue here is a valid request I believe.

**Comment by klonos** [comment]
Sorry guys, but I'm working crazy hours and have only a few hours to spare each day. Anyways, here's what I was thinking...

Perhaps the vertical tabs is not a perfect match for what we are trying to do here. I was thinking that the same model we use in the modules page for the list of modules with the addition of draggability would be mode suitable:

1. It has a checkbox for enabling/disabling.
2. It has a column to hold the filter name.
3. We can have the "more"/"less" link be a "configure" link instead that expands the table row to show/hide the config form of each filter.
4. We can have the configuration summary (currently as a "subtitle" below the tab label) displayed in a second column so that there is a simple overview.
5. We can add a drag cross to enable reordering.

What do you guys think?

**Comment by quicksketch** [comment]
I think this sounds like a pretty good plan of attack. We don't have a real pattern for this UI, but it sounds like it may work well.

**Comment by klonos** [comment]
Yeah, originally I thought that perhaps we can create some sort of "hybrid" vtabs with a checkbox on their label (to enable/disable filters), but then I figured that the checkbox can reside within the vtab. That was when I was thinking to merge only the enabling/disabling section with the config section. That's when I had a "Wait a minute" moment and said if the checkbox was not on the label, then there's room for a drag cross in there. That would allow to merge everything in one UI.

More thinking and browsing around the admin UI (plus having #1005 and the options element at the back of my head) got me considering the field UI and the modules page too. So the "hybrid" idea seemed to make more and more sense to me, but instead of the vtabs it seemed that expandable (in order to hold the config form) and draggable (in order to provide reordering functionality) table rows
was a more suitable element. Perhaps we can even style them to resemble vtabs if needed.

Another thought on the table is to use modals for config instead of expanding the table rows.

Separate issue then? ...or re-title and re-purpose this one?

**Comment by quicksketch** [comment]
> Separate issue then? ...or re-title and re-purpose this one?

Let's keep using this one. It'd be nice to see the options that we're exploring in one place, especially if this idea doesn't pan out.

**Comment by docwilmot** [comment]
This seems a really good plan. I'd vote for modals for the config. 
How would we handle disabled filters? 
- Keep them in the draggable list in the same position they were before being disabled?
- Keep them in the list but float them to the bottom and gray them out?
- Have a separate section below the list for disabled filters?
- Hide them completely and have a button/link to "Add a filter"?

**Comment by klonos** [comment]
Hmm, thinking about this and will get back with (more) comments...

In the meantime, I took the liberty to number the options you list in your comment @docwilmot so we can reference them in future comments. Hope you don't mind.

...off the top of my head, I don't particularly like option 4 unless that list gets reeeeally big. I prefer the user to be able to see the available filters without extra clicking around if possible.

**Comment by klonos** [comment]
...here's a thought:

- We ship with a sensible order for all default filters.
- Enabled ones at the top, disabled ones at the bottom (not 100% sure here because some filters might require specific ordering #1033).
- User can reorder and enable/disable at will.
- We provide a "reset to defaults" link that does the obvious.
- Once user has moved things around, we keep disabled filters in user specified place (user might change their mind or simply disabled for troubleshooting purposes).

Just thinking out loud here.

**Comment by docwilmot** [comment]
Reset to default option won't work once custom module-provided filters are enabled. There be no real "default" then. 

I think number 3 is the Drupal and Backdrop standard. Think the old blocks list, Display Suite, etc.

**Comment by docwilmot** [comment]
Is this something like what we had in mind (minus the treatment of disabled blocks)? 

Option 1

![option1](https://cloud.githubusercontent.com/assets/1042283/8763789/e46f7fa4-2d73-11e5-8e40-4fd6ee4f5fb6.PNG)


Option2

![option2](https://cloud.githubusercontent.com/assets/1042283/8763790/f7e2d4c8-2d73-11e5-88f2-4d3260beb48a.PNG)


Option3

![option3](https://cloud.githubusercontent.com/assets/1042283/8763792/0c618a3e-2d74-11e5-99c1-fbe00c3d8afd.PNG)

**Comment by docwilmot** [comment]
Other option is not have separate enable and configure links but do the enabling in the configure modal.

**Comment by MrHaroldA** [comment]
I'd suggest to keep it as close to admin/structure/types/manage/page/display as possible.

**Comment by jenlampton** [comment]
I disagree with @MrHaroldA since the manage field display page will likely not be around for long (we should be using layouts for that instead).  Let's think this through on it's own merits only.

**Comment by jenlampton** [comment]
Another option as proposed in https://www.drupal.org/node/1834682 
![combine-filter-tabs](https://cloud.githubusercontent.com/assets/397895/8994886/e1447c82-36c2-11e5-98ff-1e147c36cf05.png)

Also worth a read: https://www.drupal.org/node/1087100

**Comment by klonos** [comment]
Yes, each time I read through the commends on that issue I feel so frustrated at how things (do not) move in Drupal land. Maybe I am feeling this way because I share @quicksketch's point of view and I saw a great UI improvement idea being shot down so many times over **potential** (and may I add pre-existing) security issues that by the way we can solve here by implementing #1033.

I really do not care about the actual implementation (be it vertical tabs or dragable table rows), so long as we solve the main issue here of consolidating the filter enabling section and the filter settings section. That way users would see that there is something to configure (settings section) once they enable something in the enable/disable section. Consolidating the filter ordering section as well would be an added bonus.

So far, I like @docwilmot's option 3 (minus the spacing and alignment details). So, the order of elements I think should be: drag cross, enable/disable checkbox, filter name, configure link. I also like the idea of the configuration living inside modal dialogs.

@jenlampton: either way (dragable list or vertical tabs), you need an additional click to see the list of allowed HTML tags. Unless I'm missing something. We can have those shown in a "more" link that expands the row (à la modules page "required by" section). Still would require an additional click though :/

**Comment by docwilmot** [comment]
> in none of the options in @docwilmot's comment can you see which HTML tags are allowed without another click.

My mockup was incomplete, just to demo the order of things. The  individual draggable rows could have information about the filters below the filter name.

![filter hints](https://cloud.githubusercontent.com/assets/1042283/9014071/9f203030-3785-11e5-853c-b67626f88e7d.PNG)

And I don't really like the vertical tabs for filters to be honest; it doesnt seem like this is an appropriate use of vtabs.

> I'd suggest to keep it as close to admin/structure/types/manage/page/display as possible.

I suspect @MrHaroldA was referring to copying the UI mechanism of draggable rows above with disabled rows below.

**Comment by klonos** [comment]
> The individual draggable rows could have information about the filters below the filter name.

:+1:

**Comment by docwilmot** [comment]
I was thinking of starting to try to do this, but thought I'd get consensus first: 

tabs or draggable rows?

My first 13 votes are for draggable rows.

**Comment by quicksketch** [comment]
> The individual draggable rows could have information about the filters below the filter name.

I think this is a good idea, especially if we can include summary-like information, such as the current allowed list of tags.

I think nixing vertical tabs is also a good idea. Most of the time, settings on filters does not need to be changed, especially now that CKEditor will automatically adjust the tags based on the toolbar configuration. This will greatly simplify the overall configuration page for filters.

If we go with dialogs here, we'll probably use a new mechanism that we haven't used in the past. Right now our other dialogs (Layouts and Views) both use AJAX-saving to temporarily save configuration until the entire page is saved. I think that may be overkill here, and we'd need to completely change the way filter settings are saved to use that approach. Instead, what we'll probably want to do is leverage some of the more typical jQuery UI functionality, where we simply have a dialog open and display part of the existing page, without an AJAX request. The non-JS fallback would be identical to the page as it is today, with just all the settings for all filters being shown on the page (the configure links should also not be shown in that case).

We haven't done this particular kind of implementation before in Backdrop core. It will probably require a small amount of JS in filter.admin.js to wire it all up to the existing dialog system.

**Comment by docwilmot** [comment]
First pass, minus @quicksketch's good advice re the filter saving, which is of course the right idea, but I hadnt seen it. But at least we can evaluate this new UI.
See https://github.com/backdrop/backdrop/compare/1.x...docwilmot:filters?expand=1
Saving DOES NOT WORK. This is just a UI test.

**Comment by klonos** [comment]
...you know me, I'd love to check it out, but I still haven't sorted my local environment issues (#1072).

**Comment by klonos** [comment]
...alright. Back in business!

1. A couple stray ```dpm```s here and there (I don't have devel.module installed) - once lines were removed, I could test:

```
Fatal error: Call to undefined function dpm() in C:\www\backdrop-issue1032-comment-127066130\core\modules\filter\filter.admin.inc on line 205```
```

```
Fatal error: Call to undefined function dpm() in C:\www\backdrop-issue1032-comment-127066130\core\modules\filter\filter.theme.inc on line 105
```

2. Checking/unchecking the enable/disable checkbox would move the respective filter to/from the disabled section. "Manually" dragging and dropping would not check/uncheck the respective checkbox though:

![backdrop-new_filter_ui-dragged_filter_not_checked](https://cloud.githubusercontent.com/assets/2423362/9044157/1895861a-3a5e-11e5-8e9a-d9a92cc2f377.png)

**Comment by klonos** [comment]
3. Saving the form takes you back to the filter list page with both an error and a success message. The error is:

```Notice: Undefined index: order in filter_format_save() (line 221 of C:\www\backdrop-issue1032-comment-127066130\core\modules\filter\filter.module).```

4. Re-editing the filter shows that all filters are moved to the disabled section with their checkboxes unchecked.

**Comment by docwilmot** [comment]
Thanks @klonos. Please note the statement above:
> Saving DOES NOT WORK. This is just a UI test.

This isnt near finished yet, I know, sorry; the drag/drop worked fine earlier, I must have changed something and not rechecked. Please note my jQuery skills are laughable. 

Excuse the "Empty"; just temporary; same for the `dpm()s`

How does it work as an alternative UI for filter formats though? Is it intuitive? Friendly? Friendlier than previous?

**Comment by klonos** [comment]
I know, notice my statement:

> All in all and glitches aside (I realize this is a WIP and you did say "Saving DOES NOT WORK.") UX-wise this is looking really really great @docwilmot. Thanx!!!

**Comment by docwilmot** [comment]
Ahh we cross posted there.

Thanks for the review; I tried the checkboxes second but note that this also goes contrary to common standard which is the cross-hairs should be associated closely with what youre dragging; we're dragging the filters, not the enable checkbox. 

That, and the fact that it doesn't look at all right to me. Perhaps if having the checkbox there is unacceptable, we could opt for enabled/disabled links instead?

Agreed with the "Filter weight" header being wrong. Not sure about operations though; right now, all you can do with filters is configure them. There are no other operations.

**Comment by klonos** [comment]
...just a thought here. In order to simplify code and UI... do we actually need both ways of enabling/disabling? Should we just keep the checkboxes and ditch dragging to a "Disabled" section? In that case, should we just do what we do for Views UI (gray-out and disabled rows)?

**Comment by docwilmot** [comment]
The dragging isnt for disabling, its for relative filter weights, remember. And as per https://github.com/backdrop/backdrop-issues/issues/1032#issuecomment-121343628 if you disable, we'd still need JS to send the disabled rows elsewhere. So we'd need both the draggable code and the disabling code; why not have them work together. I dont think its that much more complication, and in fact I think its a simplification. 

And its a nice touch.

**Comment by klonos** [comment]
> Agreed with the "Filter weight" header being wrong. Not sure about operations though; right now, all you can do with filters is configure them. There are no other operations.

Yeah, I was just thinking that we could re-use the same, familiar way of the Views UI: instead of a two-click way of having to check/uncheck (or drag from/to the "disabled" section) + click "save" we could loose the entire "Enable/disable" column in favor of a "enable" button that is placed inside the "Operations" column. Once a filter is enabled, the "enable" button changes to an "edit" (or configure) drop button with "disable" as a secondary operation. Drag ordering remains as is (minus the "disabled" section thing).

Does that make sense?

**Comment by klonos** [comment]
> ...There are no other operations.

We might have a "revert"/"reset" operation in the future.

**Comment by docwilmot** [comment]
Understood; however, it makes the code more complex not less as we would still need the JS code, plus work it into dropbuttons. Unless of course we got rid of the Ajax disabling and require a page reload to disable, which is again, less good. And also enabling becomes a three step procedure (enable, drag to position, save) rather than two (drag to position, save).

> We might have a "revert"/"reset" operation in the future.

That would still fit nicely under the header "Configure" as in "revert the configuration". But I see your point. Perhaps we need a tiebreaker here.

**Comment by klonos** [comment]
> And also enabling becomes a three step procedure (enable, drag to position, save) rather than two (drag to position, save).

If the whole page was AJAXified, no save or page refresh would be required. Enable -> done. Drag -> done. Not sure if we want this though (multiple requests in the background vs a single request upon hitting save once all dragging and enabling is done).

> Perhaps we need a tiebreaker here.

I don't feel really strong about it. It's just that it seems to me that the paradigm of the Views UI seems appropriate for this UI as well to me. Plus we would establish a familiar way of UIs all over the place. I guess if not a tiebreaker per se, we definitely need a third opinion here ...or more opinions in general. @quicksketch ?? @jenlampton ?? @anyone else??

All disagreements on details aside, this is looking really good and we seem to be heading the right way though. Thanx for all you time and energy spent on this.

PS: your laughable JS skills are way better than my almost non-existent :wink:

**Comment by klonos** [comment]
Felt like we needed a more appropriate issue title here.

**Comment by docwilmot** [comment]
Next try: https://github.com/backdrop/backdrop/compare/1.x...docwilmot:filter3?expand=1

- Drag and drop fixed
- Added dialog configuring
- Fixed saving formats

**Comment by docwilmot** [comment]
Just noted if you drag to just above the "Disabled row" the checkbox stays checked. I think I know why, will fix later. Dragging anywhere else works well.

**Comment by docwilmot** [comment]
More fixes, and I think this is ready for @quicksketch to give an opinion. Only one thing I'd like is to hijack tabledrag.js to mark a row changed when a configure dialog button was clicked (and fire the message too) to give some visual indication. But I don't know how.

**Comment by quicksketch** [comment]
Hi @docwilmot, I gave this a go to see how it felt.

First off, I found the checkbox after the "Configure" link to be a little awkward. I know you'd thought about flipping them back in a previous comment (https://github.com/backdrop/backdrop-issues/issues/1032#issuecomment-122602623). After trying it out, I feel like "Option 2" would be better (with the checkbox column before configure links).

Could we change "Status" column to either "Enabled" or "Active"?

Although the Enabled/Disabled grouping pushes disables filters down, it has a very signficant downside that the default order of filters has *meaning*. By shoving all the disabled filters to the bottom, the default order is lost. Enabling a filter puts it at the bottom of the filter list instead of in its default location (which can't be reconstructed once the filters start being moved around). So although the enabled/disabled grouping has its purposes, I think it would be preferable and safer to maintain the order regardless of if a filter is enabled or disabled.

Another problem that I hadn't thought about with this approach is that validation errors are going to give us some trouble. If an error occurs inside of a dialog, the highlighted errors cannot be seen when the page reloads. You have to find the error inside of the dialog. Ideally, we'd validate the settings when the user closes the dialog, but that may require significant reworking of the form. At the very least, and AJAX request would be needed to do the validation. It should be possible however, in the same way that we validate a portion of the form when uploading a file.

I also think it would be nice if we override the theme function for the tabledrag asterisk on this page. And perhaps removed the "The changes to these filters will not be saved until the Save filters button is clicked." message (not sure about that one though).

**Comment by klonos** [comment]
> Although the Enabled/Disabled grouping pushes disables filters down, it has a very signficant downside that the default order of filters has meaning. By shoving all the disabled filters to the bottom, the default order is lost. Enabling a filter puts it at the bottom of the filter list instead of in its default location (which can't be reconstructed once the filters start being moved around). So although the enabled/disabled grouping has its purposes, I think it would be preferable and safer to maintain the order regardless of if a filter is enabled or disabled.

:+1:

> I also think it would be nice if we override the theme function for the tabledrag asterisk on this page.

Do you mean the drag cross here?

> And perhaps removed the "The changes to these filters will not be saved until the Save filters button is clicked." message (not sure about that one though).

Why do that? I think we have the same thing in other places where we have "dirty" forms fire up.

**Comment by quicksketch** [comment]
> Do you mean the drag cross here?

No I mean the asterisk that is added to the current row that has been modified. Hmm, looks like this little asterisk is just a bit more noticeable than it was prior to our new Seven theme. Let's consider that unrelated.

> Why do that? I think we have the same thing in other places where we have "dirty" forms fire up.

Yeah good point. Sometimes (as in the Views reorder dialog) we skip the reordering message. Actually, now looking at it from a 2015 stand-point, it seems odd in places like the node form. Back when tabledrag was added, pretty much any JS-related behavior was sort of unfamiliar and patterns hadn't been established. Now it's quite common to have JS-affected forms but still have save buttons. Anyway, separate issue. Both things are fine as far as this issue is concerned.

**Comment by docwilmot** [comment]
- Rolled back to single draggable table minus the disabled section.
- Rearranged so 'configure' column is last 
- Left in the functioning but unvalidating and unstyled dialog popup for configure
- Added jquery detection of disabled filters so we can dim the row text etc for visual feedback.

I hope @quicksketch can find the time to fix this before 1.2 so the filter page isn't so very busy (with all the new editor stuff). Sorry I can't finish, but JQuery will drive a man to drink.

**Comment by quicksketch** [comment]
Thanks @docwilmot. I'm hoping to come back to this before release.

**Comment by quicksketch** [comment]
Well I'm afraid we didn't get to this in time for 1.2.0. We'll have to deal with the increased complexity of the text format configuration form until 1.3.0. As this is intended to be a "polishing" release, it seems entirely appropriate anyway.

**Comment by docwilmot** [comment]
:disappointed: 

Too bad. At least its close.

**Comment by quicksketch** [comment]
> Too bad. At least its close.

It's only 4 months away, we did really well this release! :sparkles:

**Comment by klonos** [comment]
I have a gut feeling this is going to get pushed to 1.4.0 :disappointed:

**Comment by jenlampton** [comment]
As do I...  It helps if we start more than a week before release date :) whoopse. #somuchgoingon

**Comment by jenlampton** [comment]
Yep, pushed to 1.4.0. Sorry kids.

**Comment by klonos** [comment]
>  It helps if we start more than a week before release date :) ...Sorry kids.

:laughing:

**Comment by jenlampton** [comment]
No activity since last release, removing milestone.

**Comment by klonos** [comment]
...lets keep pushing to specific next milestones instead of removing. In the meantime, I'll update our roadmap page in https://backdropcms.org/roadmap (did not include PB in the list of goals for 1.4.0)

**Comment by jenlampton** [comment]
> lets keep pushing to specific next milestones...

We push one milestone, if it doesn't get done the second time, the milestone gets removed.

> did not include PB in the list of goals for 1.4.

We only include a few items on the roadmap for each release, but there are a bunch more issues tagged for each milestone. PB was a bonus!

**Comment by docwilmot** [comment]
I got close with this but gave up (a couple times).

**Comment by jenlampton** [comment]
It's a toughie one.

**Comment by klonos** [comment]
> We push one milestone, if it doesn't get done the second time, the milestone gets removed.

...my bad then. I thought we use milestones as a to-do list filter and that we pushed to the next version like we used to do for core issues in d.org.

**Comment by klonos** [comment]
Before:

![screen shot 2018-11-02 at 6 52 37 pm](https://user-images.githubusercontent.com/2423362/47900311-8faf6300-ded0-11e8-8415-1f7b3742345d.png)

After:

![screen shot 2018-11-02 at 6 51 41 pm](https://user-images.githubusercontent.com/2423362/47900314-9342ea00-ded0-11e8-8d54-2d62e4a56089.png)

**Comment by klonos** [comment]
I would like to give this a go (although my js skills are more laughable than @docwilmot's claimed his are), so I am trying to merge [@docwilmot's PR](https://github.com/backdrop/backdrop/pull/1023) into my repo: https://github.com/klonos/backdrop/pull/3 ...both PRs have conflicts, but I am not given any option to resolve them and proceed. Any pointers?

**Comment by klonos** [comment]
...OK, I've got this sorted and started taking a look. Here's a summary/heads-up by @docwilmot over Gitter:

> @klonos before you dive in consider the reason why I failed at this. its because of two suggestions from @quicksketch:
> https://github.com/backdrop/backdrop-issues/issues/1032#issuecomment-127066130
> https://github.com/backdrop/backdrop-issues/issues/1032#issuecomment-131275916
> the first is that we dont change the underlying UI for managing filters, but have the new drag/drop UI built all via javascript. that seems actually doable.
> but the second combined with the first is difficult. we'd need to pop up the existing form from the underlying page into a dialog without it having its own URL, and yet somehow make it do AJAX validation before it updates the underlying form when you save it.
> I've tried this at least 4 occasions. its hard.

Here's a thought: how about if for simplicity, we ditched the whole AJAX-config-in-a-dialog thing for now, and made the config options for each filter be a separate page/form? This would be similar to how the "Configure" action for fields in the "Manage fields" tabs for content types works (`admin/structure/types/manage/[content-type]/fields/[field-name]`), or the "Edit" operation for menu links in the "Edit links" tabs for menus (`/admin/structure/menu/item/[mlid]/edit`).

Alternatively, we could have the filter rows expand/collapse in order to reveal the config options for each filter, which would keep the filter settings in the same form.

I think that any of the above options would be much simpler than trying to figure out all that AJAX magic, and would possibly give us an MVP for 1.12.

**Comment by docwilmot** [comment]
The second option makes a lot of sense. Should have thought of that. No need for AJAX.

**Comment by jenlampton** [comment]
I like using separate configure URLs, since that would be consistent with
the rest of core. When we decide to move all these actions into modals, it
will he the same everywhere.

**Comment by docwilmot** [comment]
we have a showdown then. @quicksketch to break the tie?

**Comment by klonos** [comment]
I feel that individual filter settings should still be part of the same form; just moved from the verticals tabs into the expanding filter row. Essentially, I was planning to make the the UI look and work like the "Manage displays" UI, but nothing will be loaded via AJAX (to simplify things). So:

- the "Configure" button would expand/show the settings
- the "Update" button would hide the settings
- the "Cancel" button would revert settings to what they were when the form was loaded

All ideas in my head ATM; I have not coded anything actually. Waiting for @quicksketch and others to chime in.

**Comment by jenlampton** [comment]
As long as we don't add any new UI patterns for configuring things (we have
more than enough already) I'd love to kill the vertical tabs here :)

**Comment by klonos** [comment]
Yes, my intention exactly @jenlampton ...I was planning to keep the same UI pattern as in "Manage displays", independently of all the underlying "magic".

**Comment by jenlampton** [comment]
No movement here since Nov 2018. removing milestone, adding candidate label.

**Comment by docwilmot** [comment]
@jenlampton @klonos please have a look at the sandbox on https://github.com/backdrop/backdrop/pull/3074. I think this should satisfy everyone. The final "Save configuration" button does not yet work though (had to rush off to work and my laptop's acting funny so pushed anyway to be sure I dont lose anything).

**Comment by docwilmot** [comment]
I noticed that there is a lot of JS code that expects the allowed tags textfield to be always loaded on the page. The editor code checks the allowed tags to determine which editor buttons are available. Like if youre not allowed *bold <b>* tags you shouldnt have the bold button on the editor.

My previous PR though was loading the filter forms with AJAX, which means the textfoeld was only sometimes loaded, which would cause problems.

So a new PR loads all the filter forms, but hides them with JS.

**Comment by docwilmot** [comment]
Because we'd had some lots of discussion about the approach to this, will fix tests once this gets an initial review.

**Comment by docwilmot** [comment]
Dont bother testing yet. I've gone off in an entirely different direction, again. Should be ready by next week.

**Comment by docwilmot** [comment]
@quicksketch @herbdool @serundeputy please lend me your ear
I've started and failed at this so often its ridiculous, but this time I've got it!. I have dialogs working!

The configure button (see UI example here https://github.com/backdrop/backdrop-issues/issues/1032#issuecomment-127354846) launches the filter configuration form and the save and cancel buttons all work, for the first time. 

So this actually loads all the filter config forms on the page, in the tabledrag table, but then we hide them with JS. Then the configure button loads an instance of the form in a dialog (if you click the `filter_url` filter's "configure" button, it loads a copy of the configuration form for that filter in a dialog, but there still is the hidden form on the page). My decision here now is how to transfer data between the two form instances (the hidden one on the page and the copy that loads in the dialog)

But I've two options to update form values:

- When you close a dialog, send the values from the dialog to `Backdrop.settings` and use JS to update the underlying form. Do similar when you open the dialog to form values to the opening dialog. This is difficult as you'd need to match `input` HTML fields to data types in JQuery. (eg I found `$(input....).val(value)` doesnt work for checkboxes or selects with code I just tried. Its doable but with more robust JQuery.

- Or, when you close the dialog, save form values in tempstore, and update the underlying form with JS, and also open dialogs with values from tempstore. On the final format save, save from tempstore. This is simpler and easier. 

Would there be any objection to using tempstore here? I've failed at this so often I'm 2nd and 4th-guessing myself here.

**Comment by docwilmot** [comment]
Ps Ive asked for the peopole I asked for advice because theyre core committers; anyone else welcome to advise.

**Comment by docwilmot** [comment]
I've got a PR that accomplishes the goal and passes tests. Here's how it works:
- The page loads a tabledrag table with the settings forms for each filter hidden by JS
- On final save, the form reads the values in these hidden settings forms to save the filters and formats
- On clicking the Configure button, this loads a copy of the settings form in a dialog, by calling the same function that creates the hidden settings forms.
- This means the dialogs load with default values, then on closing the dialog, it just updates the underlying form with the dialog values.
- Leading to needing some JS to swap form field values between the two sets of settings forms. So on saving a dialog, the JS reads the fields in the dialog, finds the corresponding fields in the underlying settings form, and changes the underlying field values to what was in the dialog. But in case a user closes the dialog then changes their mind, on *re-opening* a dialog, we also read the fields from the underlying form to re-establish the value the user initially set.

There very likely is a better way to do this, but I have found one. Thoughts welcome though.

**Comment by klonos** [comment]
Initial feedback (pending more thorough review): this looks and works beautifully!! ...thank you for all your work on this @docwilmot 👍 

Things that annoyed me UX-wise:

1. Dialogs are too wide
1. Opening a dialog makes it so that the page "jumps" to the top
1. Saving the form returns me to the text formats index. I realize that this may be desired(?), but it annoyed me while texting
1. I've removed the `<strong>` tag from the list of allowed tags, but even after saving the form, the "B" button was in the "Available buttons" listing. It should either be hidden, or "disabled" (grayed out + undraggable)
1. With the `<strong>` tag still removed (from previous testing step), I have disabled the "Limit allowed HTML tags" filter -> then dragged the "B" button from the available buttons to the active toolbar -> saved the form -> edit the filter again -> find the "B" button removed from the active toolbar, and moved back into the list of available buttons 👎 ...I would expect that if a filter is disabled, whatever effects it would otherwise have on the toolbar are cancelled.

**Comment by docwilmot** [comment]
1 can fix
2 I noticed, definitely annoying, will try to fix
3 Is default behaviour
4 and 5 I forgot to look into that association between the tags and the buttons, will do

Thanks

**Comment by BWPanda** [comment]
I haven't looked at the code or tried the PR sandbox; my comments are based solely on @docwilmot's above explanations...

It seems to me that it's not ideal having a bunch of forms load on the page, then hide them with JS, then load a *copy* of one of those forms in a dialog, then have to copy values between the hidden and dialog forms. I'm not sure if Backdrop has a policy regarding how usable a site is with JS disabled...? But having all the forms show up if so can't be pretty/usable. How does Views do it? It has plenty of forms popup in dialogs/modals. Can we use the same thing here?

Ideally, no hidden forms would load initially. Only when you click Configure would the appropriate form be loaded (for the first time) in the popup. And since it's the main copy of the form, you wouldn't need to copy values to and fro.
Mind you, this is coming from the perspective of someone who wouldn't have a clue how to code that, so it may very well not be possible/easily doable (so take this with a grain of salt). But that'd be how I'd imagine we'd like it to work if at all possible...

**Comment by docwilmot** [comment]
@BWPanda actually, in current Backdrop, all the forms show up on the page as well, but in vertical fieldsets. 

The issue with the default Backdrop now, version 15.0.0, to save you reading the issue, is that we have on the format configuration pages:
- a checkbox list to set which filters are enabled
- a separate draggable table to set the order of the enabled ones
- and a vertical fieldset to keep the settings form for those which have settings
This is obviously quite redundant, and this PR simply aims to combine all three of those into one: a draggable table with an enabling checkbox and the settings form within the draggable table.

As noted already, the idea of having the settings form load within the table is similar to how field display settings work (see https://github.com/backdrop/backdrop-issues/issues/1032#issuecomment-126478886); note I also have a PR (linked in the first issue comment) that does it this way (https://github.com/backdrop/backdrop/pull/3077). But we felt that loading the settings form in dialogs would be neater.

We could yes load the settings form entirely in dialogs, but then if JS is disabled we'd have a problem. Other forms (like layouts) that use dialog have a normal menu fallback, where the form that would have loaded in the dialog would have its own page, and with disabled JS, we'd just redirect to that page. @quicksketch at some point in this very long issue didnt want this, as it would change how filters are saved, he wanted the settings forms to load on the format page, even if hidden. See https://github.com/backdrop/backdrop-issues/issues/1032#issuecomment-127066130.

Having settings forms have their own URLs would have made this issue over and done about 4 years ago, but I couldnt manage until now, with this kludgy idea. 

Maybe we could revisit @quicksketch's reservations here. Actually been fixated on that without considering we've moved on a bit in Backdrop, and maybe its time to do just that.

But if we stick to this idea of having the settings form hidden on the format page, we need some way to be able to load that same form in a dialog. We need the submit handler to also save final values from the hidden form (in case JS is disabled). So then we must either use JS to move the entire form HTML from dialog to hidden form, back and forth, or move just the values, back and forth.

And here we are.

**Comment by BWPanda** [comment]
I think having this work the same as the rest of Backdrop (separate URLs for non-JS, those URLs opened in a popup with JS) would be ideal. But I'd be interested to hear from @quicksketch to see if his objections to that have changed over the years...

If we do decide to stick with the hidden/popups forms though, would it be possible to do something like in the examples seen here: https://mdbootstrap.com/docs/jquery/modals/forms/ ?
Basically you have an HTML form on a page (hidden with CSS/JS/whatever), and then when you click to open the popup, it populates the popup with the HTML from the element you specify (e.g. `data-target="#modalLoginForm"`). So there's no copying back and forth. You just use the existing form. There's plenty of examples of this across the web, that's just the first I came across when searching. Hope that explains my point...

**Comment by docwilmot** [comment]
Agreed with everything @BWPanda 
Just one other consideration; I had considered the alternatives of popping up the form in a non-Backdrop-AJAX dialog, but we also need validation functions to run when we hit the dialog submit button. I wouldnt know how to do that (we'd need to run the validate function and reload the dialog with the error message in the dialog; how to do that with a Bootstrap dialog for example I wouldnt know).
So in summary then, I think wait for @quicksketch

**Comment by docwilmot** [comment]
New PR up and passing!

**Comment by stpaultim** [comment]
Sandbox doesn't seem to be working.

**Comment by klonos** [comment]
Yeah ^^ ...assuming that the latest PR is https://github.com/backdrop/backdrop/pull/3096

**Comment by docwilmot** [comment]
Rebooted and working.

**Comment by docwilmot** [comment]
Ready for review

**Comment by jenlampton** [comment]
I'm so sorry I forgot to add my comment, but yes -- this is a real improvement!  

It looks like the HTML tags in the `Allowed HTML tags` field don't seem to be updating anymore. Before the PR, these would update as buttons were dragged in and out of the `Active toolbar` area. 

Minor design suggestions:

1) The `Configure` button should use the drop-button styles like it does in other interfaces:

Formats:
<img width="1223" alt="Screen Shot 2020-04-16 at 1 52 17 PM" src="https://user-images.githubusercontent.com/397895/79505710-a1fc5e80-7fe9-11ea-80b1-00e7b87ab88e.png">

Field UI:
<img width="1199" alt="Screen Shot 2020-04-16 at 1 52 26 PM" src="https://user-images.githubusercontent.com/397895/79505694-9dd04100-7fe9-11ea-8c16-44b69f8079f2.png">

**Comment by quicksketch** [comment]
I posted a code review at https://github.com/backdrop/backdrop/pull/3096#pullrequestreview-400440896

This looks really great! :100:

My suggestions _might_ help with updating the allowed HTML values per Jen's feedback.

> The Configure button should use the drop-button styles like it does in other interfaces:

That may need to be a follow-up. Unlike other forms where the drop-button contains links, these Configure buttons are part of a multi-part form, and must be HTML `<input>` elements for the form to submit. We'll need to update Seven to accommodate this as theme-specific changes.

**Comment by jenlampton** [comment]
> That may need to be a follow-up. 

That's reasonable. I've created https://github.com/backdrop/backdrop-issues/issues/4378 and tagged it with the milestone so we don't forget :)

**Comment by quicksketch** [comment]
I'm going to work on addressing the feedback from the latest code review.

**Comment by quicksketch** [comment]
New pull request up at https://github.com/backdrop/backdrop/pull/3131

I reworked the updating of the hidden "filtered_html" element and fixed the interaction between the configuration of CKEditor's buttons and the tag list. It should now work consistently when you drag in and out buttons like Image or Strike-Through.

I also encountered a number of problems when creating a _new_ text format, some PHP errors would be thrown when configuring filters before giving it a name.

We need a code review and some additional testing on this.

**Comment by jenlampton** [comment]
I'm testing the latest PR, thoughts:

* In my testing all the tags I needed based on my editor configuration were added to the `Limit allowed HTML tags` filter configuration.
* I **really like** the new text that explains why tags have been added:
`Based on the text editor configuration, these tags have automatically been added:...`

These should maybe be a follow-ups: 
* when creating a new text format I struggled to determine how to turn on a filter, since the "Enabled" column wasn't on the left hand side of the filter name. 
* (deleted, on second thought)

**Comment by quicksketch** [comment]
I also thought that the checkbox might be more expected in the left-hand column, similar to enabling modules or any of the views bulk operations forms (i.e. admin/content).

We can change the dialog submit button very easily, it's just the label on the button, no need to adjust any JS.

**Comment by jenlampton** [comment]
> We can change the dialog submit button very easily, it's just the label on the button, no need to adjust any JS.

Ah! Maybe it's the tests that will need to be updated?

**Comment by jenlampton** [comment]
I added a few commits to the PR to update the CSS (adding RTL styles for filter UI) and updated the docs in one place to indicate Filters are stored as config, not in the database anymore, but as long as we can get someone else's eyes on that before it's merged, this issue is RTBC!

**Comment by quicksketch** [comment]
Changed the dialog button label from the generic "Submit" to the slightly more specific "Update".

**Comment by quicksketch** [comment]
> when creating a new text format I struggled to determine how to turn on a filter, since the "Enabled" column wasn't on the left hand side of the filter name.

Let's consider moving the checkbox for enabling as part of the follow-up we already have to style the Configure buttons over at #4378.

**Comment by jenlampton** [comment]
> Let's consider moving the checkbox for enabling as part of the follow-up we already have to style the Configure buttons over at #4378.

Great, adding it there now :)

**Comment by quicksketch** [comment]
Merged https://github.com/backdrop/backdrop/pull/3131 into 1.x for 1.16.0.

Thanks everyone for the extensive feedback and testing!

HUGE THANK YOU to @docwilmot for spearheading this and making not one but TWO entire implementations to get this done. Really happy with how this has turned out.
